### PR TITLE
Tests to routers

### PR DESCRIPTION
**Pull request by avivsarig**
Created 4 test files - 1 for each router:
beers, orders, stock and users.

added 2 utility functions in test folder:
- apply_filter - to simulate the filtering process
- payload_to_string - for db.execute() insertion

### CODE REVIEW

File: backend/models.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ class User(BaseModel):
     email = CharField(60, unique=True)
     password = CharField(60)
     address = CharField(60, null=True)
-    phone = CharField(12, null=True)
+    phone = CharField(20, null=True)
```
Comment (itraviv) [inline backend/models.py:27]: what about the migration?

File: backend/routers/beers.py [hunk 2]
```diff
@@ -62,7 +62,8 @@ async def delete_beer(beer_id):
         raise HTTPException(status_code=404, detail="Beer not found")
     else:
         with db.atomic():
```
Comment (itraviv) [inline backend/routers/beers.py:64]: this is redundant

File: backend/routers/orders.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 async def get_orders(request: Request):
     query = Order.select()
     filters = query_to_filters(request["query_string"])
+    print(filters)
```
Comment (itraviv) [inline backend/routers/orders.py:16]: remove debug print

File: backend/routers/orders.py [hunk 4]
```diff
@@ -35,37 +36,31 @@ async def get_orders(request: Request):
...
+        raise HTTPException(status_code=code, detail=message)
 
 
 @router.post("/")
 async def create_order(request: Request):
     body = await request.json()
-    with db.atomic():
-        try:
+    try:
+        with db.atomic():
```
Comment (itraviv) [inline backend/routers/orders.py:57]: redundant - no need for `atomic`

File: backend/routers/orders.py [hunk 5]
```diff
@@ -76,7 +71,12 @@ async def delete_order(order_id):
             raise HTTPException(status_code=404, detail="Order not found")
         else:
             with db.atomic():
```
Comment (itraviv) [inline backend/routers/orders.py:73]: same here

File: backend/routers/stock.py [hunk 6]
```diff
@@ -45,13 +44,16 @@ async def get_stock_by_id(stock_id):
 @router.post("/")
 async def create_stock(request: Request):
     body = await request.json()
+    if type(body.get("qty_in_stock")) == float:
```
Comment (itraviv) [inline backend/routers/stock.py:47]: use `pydantic`

File: backend/routers/stock.py [hunk 7]
```diff
@@ -68,6 +70,8 @@ async def delete_stock(stock_id):
 @router.put("/{stock_id}")
 async def update_stock(request: Request, stock_id: int):
     body = await request.json()
+    if type(body.get("qty_in_stock")) == float:
```
Comment (itraviv) [inline backend/routers/stock.py:73]: fix this using pydantic

File: backend/routers/users.py [hunk 8]
```diff
@@ -67,14 +70,21 @@ async def delete_user(user_id):
...
-    body = await request.json()
     try:
         query = User.select().where(User.id == user_id)
         if not query.exists():
             raise HTTPException(status_code=404, detail="User not found\n")
-        else:
-            with db.atomic():
-                User.update(**body).where(User.id == user_id).execute()
+
+        user_from_db = query.get().__dict__["__data__"]
```
Comment (itraviv) [inline backend/routers/users.py:78]: why?

File: backend/routers/users.py [hunk 9]
```diff
@@ -67,14 +70,21 @@ async def delete_user(user_id):
...
+        body = await request.json()
+
+        for key in user_from_db.keys():
+            if key not in body.keys():
+                body[key] = user_from_db[key]
+
+        with db.atomic():
```
Comment (itraviv) [inline backend/routers/users.py:85]: not need for transaction

File: backend/routers/users.py (hunk 8 above)
Comment (itraviv) [reply backend/routers/users.py:78]: user <User.1>
user.phone = '123123'
user.save()

File: backend/routers/users.py [hunk 10]
```diff
@@ -67,14 +70,21 @@ async def delete_user(user_id):
...
         if not query.exists():
             raise HTTPException(status_code=404, detail="User not found\n")
-        else:
-            with db.atomic():
-                User.update(**body).where(User.id == user_id).execute()
+
+        user_from_db = query.get().__dict__["__data__"]
+        body = await request.json()
+
+        for key in user_from_db.keys():
```
Comment (itraviv) [inline backend/routers/users.py:81]: refactor this, you don't need it

File: backend/models.py [hunk 11]
```diff
@@ -46,7 +46,7 @@ class Stock(BaseModel):
     id = BigAutoField()
     beer_id = ForeignKeyField(Beer, unique=True)
     date_of_arrival = DateTimeField()
-    qty_in_stock = IntegerField()
+    qty_in_stock = IntegerField(constraints=[Check("qty_in_stock >= 0")])
```
Comment (itraviv) [inline backend/models.py:49]: need migration

File: backend/tests/test_beers_router.py [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+
+
+def create_payload(name=None, style=None, abv=None, price=None):
+    return {key: value for key, value in locals().items() if value is not None}
+
+
+@pytest.fixture(scope="function")
+def clean_db():
+    with db.atomic():
+        db.rollback()
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:45]: don't need rollback here, also atomic is not needed

File: backend/tests/test_beers_router.py [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+    password="iLoveTesting",
+    host="localhost",
+    port=5432,
+)
+
+client = TestClient(app)
+
+
+def setup_module():
+    db.bind([Beer])
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:29]: why not on all models?

File: backend/tests/test_beers_router.py (hunk 12 above)
Comment (itraviv) [reply backend/tests/test_beers_router.py:45]: i would suggest using a per-test DB - try to achieve that

File: backend/tests/test_beers_router.py [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+@pytest.fixture(scope="function")
+def clean_db():
+    with db.atomic():
+        db.rollback()
+        db.execute_sql("TRUNCATE TABLE beers RESTART IDENTITY CASCADE;")
+    yield
+
+
+class Test_create_beer:
+    @pytest.mark.parametrize(
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:51]: NICE!

File: backend/tests/test_beers_router.py [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+        "name, style, abv, price",
+        [("test_beer1", None, 1.1, 1.11), ("test_beer2", "test_style2", 2.2, 2.22)],
+    )
+    def test_create_beer_content(self, clean_db, name, style, abv, price):
+        payload = create_payload(name, style, abv, price)
+        client.post("/beers/", json=payload)
+        query = {
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:58]: there should be a way more simpler and coherent way to write this.

File: backend/tests/test_beers_router.py [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+        assert response.status_code == 400
+
+    def test_create_fail_invalid_price(self, clean_db):
+        response = client.post(
+            "/beers/", json=create_payload("test_fail_price", "test_style", 1.0, -1.11)
+        )
+        assert response.status_code == 400
+
+
+class Test_delete_beer:
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:105]: class names are CamelCase (pep8)

Comment (itraviv) [reply backend/tests/test_beers_router.py:105]: use `flake8` to lint your code

File: backend/tests/test_beers_router.py [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+    def test_delete(self, clean_db):
+        with db.atomic():
+            id = db.execute_sql(
+                "INSERT INTO beers (name, style, abv, price) VALUES ('test_delete', 'test_style', 5.0, 5.55) RETURNING id;"
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:109]: NICE `returning` is some advanced feature.. great

File: backend/tests/test_beers_router.py [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+            ).fetchall()[0][0]
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:110]: huh?

File: backend/tests/test_beers_router.py [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+        client.delete(f"/beers/{id}")
+        assert (
+            len(db.execute_sql(f"SELECT 1 FROM beers WHERE id = {id}").fetchall()) == 0
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:113]: why not `Beers.select().where(Beers.id == id)` - you are mixing syntaxes.
also there is SQL injection here.

in general - Never use pure SQL on user input. You will always need input sanitation which is always not easy.

File: backend/tests/test_beers_router.py (hunk 17 above)
Comment (itraviv) [reply backend/tests/test_beers_router.py:109]: although, i would write this down in a peewee way

File: backend/tests/test_beers_router.py [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+        [
+            ("test_beer1", None, None, None),
+            (None, "test_style2", None, None),
+            (None, None, 4.0, None),
+            (None, None, None, 14.99),
+        ],
+    )
+    def test_update_content(self, clean_db, name, style, abv, price):
+        with db.atomic():
+            id = db.execute_sql(
```
Comment (itraviv) [inline backend/tests/test_beers_router.py:279]: never use a parameter called `id`  - you are shadowing the function `id`

Comment (itraviv) [reply backend/tests/test_beers_router.py:279]: same goes for all `builtins`

File: backend/models.py (hunk 1 above)
Comment (avivsarig) [reply backend/models.py:27]: Will be included in next sprint

File: backend/tests/test_beers_router.py (hunk 13 above)
Comment (avivsarig) [reply backend/tests/test_beers_router.py:29]: Why would I need it if no other model is relevant to these test?
Of course I can add them, but why?

**Pull request closed without merging**